File: examples/ch04/IntParse.hs
-- Idea courtesy of William Lee Irwin.

{-- snippet type --}
import Data.Char (ord)

asInt :: String -> Int
{-- /snippet type --}

{-- snippet loop --}
loop :: Int -> String -> Int

asInt xs = loop 0 xs
{-- /snippet loop --}

{-- snippet base --}
loop acc [] = acc
{-- /snippet base --}

{-- snippet inductive --}
loop acc (x:xs) = let acc' = acc * 10 + ord x - ord '0'
                  in loop acc' xs
{-- /snippet inductive --}

File: examples/ch05/intparse.c
#include <stdio.h>

/** snippet as_int */
int as_int(char *str)
{
    int acc;

    for (acc = 0; *str != '\0'; str++) {
	acc = acc * 10 + *str - '0';
    }

    return acc;
}
/** /snippet as_int */ 

int main(int argc, char **argv)
{
    int i;
    
    for (i = 1; i < argc; i++) {
	printf("%d\n", as_int(argv[i]));
    }
}
